crates/pg_statement_splitter/src/parser.rs

```diff
@@ -33,164 +37,94 @@ pub struct Parse {
 }
 
 impl Parser {
-    pub fn new(tokens: Vec<Token>) -> Self {
+    pub fn new(sql: &str) -> Self {
+        // we dont care about whitespace tokens, except for double newlines
+        // to make everything simpler, we just filter them out
+        // the token holds the text range, so we dont need to worry about that
+        let tokens = lex(sql)
+            .iter()
+            .filter(|t| {
+                return !WHITESPACE_TOKENS.contains(&t.kind)
+                    || (t.kind == SyntaxKind::Newline && t.text.chars().count() > 1);
+            })
+            .rev()
+            .cloned()
```

Comment:
nit: `into_iter()` instead of `.cloned()` ?

I'm not sure, but if I understand it correctly, we could skip some allocations

Reply:
this will make `tokens` a reference. but something feels off about this implementation for sure. I would rather use a pointer index instead of popping from the vector. I was struggling to implement this nicely because I wanted the pointer to be an `usize` no matter what. but we need to start the pointer at `-1`, or always subtract 1 when accessing the vector. but I might give it another go.

Reply:
I refactored this a bit. see below :)